### PR TITLE
chore(*) use string.byte to optimize dns client.lua

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

dns.client.lua use many string.sub to manipulate string,
We can use string.byte to optimize check first/last char,
It is better than string.sub.

### Full changelog

* use string.byte instead of string.sub

### CODE REVIEW

**Review by bungle — APPROVED** [review]

**Comment by javierguerragiraldez** [comment]
i want to note a couple things here:

1. i do like this kind of tweaks, i think string.byte() is a better expression of "i want to check this ASCII char", so readability is improved.

2. but...
  2.1. the escape analysis can statically determine that the `string:sub()` result never goes anywhere, so no new string is allocated/copied/interned
  2.2. the compared strings are 1byte long each, so it's just a byte compare, not a hashed string function call
  2.3. constant propagation would compile `string.byte('x')` into a constant itself, so the upvalue isn't really needed either
  2.4. upvalues aren't locals, they're GC objects by themselves, and using them requires at least one extra reference (still much better than the hash table access for globals, but far more expensive than real locals)

try:
```
luajit -j dump -e "s='abc'; k=0; for _ = 0, 60 do for i=1, #s do if s:sub(i,i)=='A' then k=k+1 end end end"
```
and see in the IR code how there's no new string interning, no string comparison (just a constant compare with +65 (ASCII 'A')), and no call to string.sub (just a check to verify it's still the function it was supposed to be)


TL;DR:  yes, i like the code better like this, i think it's more readable.  but no, it won't be faster

**Comment by chronolaw** [comment]
> i want to note a couple things here:
> 
> 1. i do like this kind of tweaks, i think string.byte() is a better expression of "i want to check this ASCII char", so readability is improved.
> 2. but...
>    2.1. the escape analysis can statically determine that the `string:sub()` result never goes anywhere, so no new string is allocated/copied/interned
>    2.2. the compared strings are 1byte long each, so it's just a byte compare, not a hashed string function call
>    2.3. constant propagation would compile `string.byte('x')` into a constant itself, so the upvalue isn't really needed either
>    2.4. upvalues aren't locals, they're GC objects by themselves, and using them requires at least one extra reference (still much better than the hash table access for globals, but far more expensive than real locals)
> 
> try:
> 
> ```
> luajit -j dump -e "s='abc'; k=0; for _ = 0, 60 do for i=1, #s do if s:sub(i,i)=='A' then k=k+1 end end end"
> ```
> 
> and see in the IR code how there's no new string interning, no string comparison (just a constant compare with +65 (ASCII 'A')), and no call to string.sub (just a check to verify it's still the function it was supposed to be)
> 
> TL;DR: yes, i like the code better like this, i think it's more readable. but no, it won't be faster

Thanks for your comment, I learned a lots.
